### ISSUE TITLE
auto bond: different from cpptraj

### ISSUE THREAD

**Issue by hainm** [ParmEd/ParmEd#688]
:D just tested ParmEd's new PR

parmed gives 4880 bond vs. 4879 cpptraj.

file: https://github.com/hainm/amber_things/tree/master/things/parmed/autobond

```python
import pytraj as pt, parmed as pmd

fn = '2pia_2pia1423_0001.pdb'

traj = pt.load(fn)
parm = pmd.load_file(fn)

print("parmed")
print(len(parm.bonds))

print("cpptraj")
traj.top.summary()
```

get bond_indices
```python
traj.top.bond_indices
```

**Comment by hainm** [comment]
by the way: for 4828 atoms, the speed is quite fast.

**Comment by swails** [comment]
The algorithms that ParmEd and cpptraj use are very different.  cpptraj does a simple distance-based bond assignment approach where atoms have to be within a certain distance (for specific element pairs) in order to count as a bond.

By contrast, tleap takes an approach that is entirely template-based.  Bonds are assigned from templates where residue and atom names are assigned by name, and all bonds are explicitly defined inside the templates themselves, with a head and tail attribute telling them how to interconnect residues.

ParmEd takes a hybrid approach.  I've specified a "typical" set of standard biomolecular residues in `parmed.modeller.StandardBiomolecularResidues` and assign all bonds from those templates where atom names and residue names match.  Then the head and tail attributes are used to stitch them together.  Any atoms or residues that were *not* in the template are stored for later processing, and bonds between those atoms are assigned based on the same distance criteria that cpptraj uses, with a couple caveats:

1. Any atom whose residue is in a template but who is not in a template itself will only attempt to bond with other unassigned atoms and with atoms inside its residue.  In this way, mis-named H atoms are properly bonded within its residue wherever that may happen (as it does in 2pia_2pia1423_0001.pdb, for example, where the H atom names are PDB v2 instead of v3).
2. Cysteine SG atoms not bonded to 2 other atoms (a C and an H) look for other atoms to bond to, since CYS is often a place where post-translational modifications are made.

So in particular, ParmEd will disagree with cpptraj in a few instances:

* Where two atoms that, in a standard residue, are bonded together but whose distances are a little too far apart to fall within cpptraj's criteria will be identified by ParmEd but not cpptraj
* Gaps where there is no TER and no change in chain will result in a bond crossing that gap (this is the same thing tleap does)
* Atoms within standard residues that are a little too close to one another (imagine a methyl group in an average PDB where all 3 Hs are basically on top of each other) will grow extra bonds in cpptraj (I think), since each H will be close enough to register bonds with each other.  Since they are assigned via template in ParmEd, the canonical set of bonds will be applied and no more

In the case of `2pia_2pia1423_0001.pdb` here, I stripped out the rosetta information at the end, removed hydrogens, and created a prmtop using ff14SB.  This is what I got:

```Python
>>> import parmed as pmd
>>> pmd.load_file('prmtop', 'inpcrd')
<AmberParm 4828 atoms; 315 residues; 4880 bonds; parametrized>
>>> pmd.load_file('2pia_2pia1423_0001.pdb')
<Structure 4828 atoms; 315 residues; 4880 bonds; NOT parametrized>
```

And this is what I *want* -- I *want* ParmEd to assign a topology to a PDB file the same way that tleap (and related tools) would.  For cpptraj this is not nearly as important.

> by the way: for 4828 atoms, the speed is quite fast.

This is a middle-ground test case.  ParmEd will assign half of the bonds from templates, and all inter-residue interconnects from templates as well.  All H atoms will need to have their bonds assigned, and that assignment is done by distance.  So the pairlist is generated between 2381 H atoms, most of which will have at most 2 or 3 atoms in its pairlist (none of which will be close enough to bond).  Then it cycles through the 20 atoms inside its residue and finds its bond partner by distance.  If the PDB obeyed PDB v3, the bond search would be quite a bit faster.

The real test comes when you have a lipid or carbohydrate system where tens of thousands of atoms all have to have bonds assigned by distance.  In one such case, Travis couldn't finish the bond search within the 10 minutes allowed to any single test :).  In that case, the pairlist I added drops the bond search time from >10 minutes to ~1 minute in pathological cases (and it scales as O(N) rather than the naive O(N^2) that the simple all-pairs-search scales).  For what I need it for, the pairlist builder is very good.  It's fast to build and results in quite compact pairlists for my needs.

**Comment by swails** [comment]
TL;DR -- ParmEd gets the bond count right.

**Comment by hainm** [comment]
cool. should ping @drroe here. Do you think ParmEd and cpptraj should be matched in this aspect?

**Comment by swails** [comment]
> Do you think ParmEd and cpptraj should be matched in this aspect?

No.  Inefficient use of Dan's time.  <-- TL;DR.  Longer explanation below.

----------

There's a lot of functionality in ParmEd that's not in cpptraj (and wouldn't make sense to put in cpptraj) that made it easier for me to do:

* OFF library parsing <-- probably the big one
* `ResidueTemplate` and "standard" residue databases/mappings (e.g., ParmEd knows that HIE, HID, and HIP are all HIS)
* Raw performance that is prohibitively slow for simple distance-based bond assignment (an algo that takes 1 second in C++ takes Python 2 minutes), so I needed something a lot faster
* ParmEd is topology-centric, while cpptraj is trajectory-centric

Most cpptraj users *don't* use PDB files as topologies, so the effort to fix some edge cases like this wouldn't even impact the majority of users.  And what does "getting this right" do for cpptraj?  *Maybe* fix some (*incredibly* rare) imaging artifacts. I challenge you to find a non-contrived example where ParmEd's approach would fix cpptraj

The main reason I took such a "sophisticated" approach to bond assignment was to make bond searching faster, and it also helped out some other aspects of ParmEd that doesn't affect cpptraj.

OTOH, pytraj is a lot more integrated into other projects in the ecosystem, so it's not unreasonable that some uses of pytraj would benefit from ParmEd's approach.  But in those cases, you can *use* ParmEd's approach directly through ParmEd.

**Comment by hainm** [comment]
> But in those cases, you can use ParmEd's approach directly through ParmEd.

True

**Comment by swails** [comment]
I think this can be closed?

**Comment by hainm** [comment]
example :D 

```python
In [2]: len(parm.bonds)
Out[2]: 4880

In [3]: parm.save('test.mol2')

In [4]: traj = pt.load('test.mol2')

In [5]: len(traj.top.bond_indices)
Out[5]: 4880
```

**Comment by swails** [comment]
This is a good thing, right?

**Comment by hainm** [comment]
Sure it is.

**Comment by drroe** [comment]
> Atoms within standard residues that are a little too close to one another (imagine a methyl group in an average PDB where all 3 Hs are basically on top of each other) will grow extra bonds in cpptraj (I think), since each H will be close enough to register bonds with each other.

Actually cpptraj will not try to bond any H atom more than once. It knows at least that much about chemistry :-)

**Comment by swails** [comment]
That's more chemistry than ParmEd knows :)
